librad/src/id.rs

```diff
@@ -214,8 +284,109 @@ pub trait Entity {
             Err(Error::SignatureVerificationFailed)
         }
     }
+
+    fn check_validity(&self, resolver: &impl Resolver<User>) -> Result<(), Error> {
+        let mut keys = HashSet::<PublicKey>::from_iter(self.keys());
+        let mut users = HashSet::<RadicleUri>::from_iter(self.certifiers());
+
+        for (k, s) in self.signatures() {
+            self.check_signature(k, &s.by, &s.sig, resolver)?;
+            match &s.by {
+                Signatory::OwnedKey => {
+                    keys.remove(k);
+                },
+                Signatory::User(user) => {
+                    users.remove(&user);
+                },
+            }
+        }
+        if keys.len() > 0 || users.len() > 0 {
+            Err(Error::SignatureMissing)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn is_valid(&self, resolver: &impl Resolver<User>) -> bool {
+        self.check_validity(resolver).is_ok()
+    }
+
+    fn check_update(&self, previous: &Self) -> Result<(), UpdateVerificationError> {
+        if self.revision() <= previous.revision() {
+            return Err(UpdateVerificationError::NonMonotonicRevision);
+        }
+
+        let retained_keys = self.keys().filter(|k| previous.has_key(k)).count();
+        let total_keys = self.keys_count();
+        let added_keys = total_keys - retained_keys;
+        let removed_keys = previous.keys_count() - retained_keys;
+        let quorum_keys = total_keys / 2;
+
+        if added_keys > quorum_keys {
+            return Err(UpdateVerificationError::NoCurrentQuorum);
+        } else if removed_keys > quorum_keys {
+            return Err(UpdateVerificationError::NoPreviousQuorum);
+        }
+
+        let retained_certifiers = self
```

Comment:
These two look identical in form, can we write a function that handles this pattern?